### PR TITLE
227 automatic handling of synthetic stage axes gui should only display available axes

### LINKED ISSUE #227: Automatic Handling of Synthetic Stage Axes, GUI should only display available Axes

### PR DESCRIPTION
**Pull request by AdvancedImagingUTSW**
I want to see how this looks on a Windows machine.

### CODE REVIEW

**Comment by AdvancedImagingUTSW** [comment]
@zacsimile - Want your opinion. Do you think it would be better to destroy the axes if they are not available, or disable them?

**Comment by AdvancedImagingUTSW** [comment]
This is what it looks like on a Windows machine... If everyone likes how this looks, I can proceed and fix the tests, and also make sure that the spinbox is also disabled. I prefer this over deleting the buttons, since things start to look pretty odd when they go missing (also requires more code to repopulate it if we switch to non-synthetic stages).
image

File: src/navigate/model/microscope.py [hunk 1]
```diff
@@ -276,23 +276,24 @@ def __init__(
             if device_ref_name not in devices_dict["stages"]:
                 logger.debug("stage has not been loaded!")
                 raise Exception(
-                    "Stage not found. \n"
+                    f"Stage not found. "
```
Comment (zacsimile) [inline src/navigate/model/microscope.py]: Why the f string?

File: src/navigate/controller/sub_controllers/stage_controller.py [hunk 2]
```diff
@@ -167,6 +172,8 @@ def stage_key_press(self, event):
     def initialize(self):
         """Initialize the Stage limits of steps and positions."""
         config = self.parent_controller.configuration_controller
+        self.disable_synthetic_stages(config)
```
Comment (zacsimile) [inline src/navigate/controller/sub_controllers/stage_controller.py:175]: This probably also needs to be called when switching microscopes.

File: src/navigate/config/configuration.yaml [hunk 3]
```diff
@@ -121,7 +121,7 @@ microscopes:
       hardware:
         -
           name: stage
-          type: PI
+          type: syntheticstage
```
Comment (zacsimile) [inline src/navigate/config/configuration.yaml]: We probably shouldn't update `configuration.yaml` in the main branch.

File: src/navigate/model/microscope.py (hunk 1 above)
Comment (AdvancedImagingUTSW) [reply src/navigate/model/microscope.py]: I think my goal was to provide a more informative Exception and state which stage was not found. I'll go back and do that. Given that this is one of the fragile break points in getting the code set up, I want all feedback to be as useful as possible in diagnosing the problem.

Comment (AdvancedImagingUTSW) [reply src/navigate/model/microscope.py]: If you look at the last line of this string, there is a variable for the f-string, but I see it is formatted incorrectly. I will fix it.

File: src/navigate/config/configuration.yaml (hunk 3 above)
Comment (AdvancedImagingUTSW) [reply src/navigate/config/configuration.yaml]: I accidentally did this while testing it.

File: src/navigate/controller/sub_controllers/stage_controller.py (hunk 2 above)
Comment (AdvancedImagingUTSW) [reply src/navigate/controller/sub_controllers/stage_controller.py:175]: I created a second `microscope` on the mesoSPIM with only synthetic stages. 

The first microscope is still defined with the ASI stages:
image

Here is the one with synthetic stages:
image

And, when I switch back to the original mesoSPIM, all of the stage axes become active again. Nice call @zacsimile

**Comment by AdvancedImagingUTSW** [comment]
I currently think that this is functional. If everyone likes the addition, I will move to add a unit test or two and convert it to a final PR. I just made one more push to make sure that stages can be read in both as a DictProxy and as a ListProxy. Specifically, the difference between this configuraiton file:
```
    stage:
      hardware:
        name: stage
        type: SyntheticStage
        serial_number: 0000
        axes: [x, y, z, f, theta]
        feedback_alignment: [88, 91, 91, 91, 85]
        axes_mapping: [X, V, Z, Y, T]
        volts_per_micron: None
        axes_channels: None
        max: 5
        min: 0
```
And this configuration file:
```
    stage:
      -
        hardware:
          name: stage
          type: SyntheticStage
          serial_number: 0000
          axes: [x, y, z, f, theta]
          feedback_alignment: [88, 91, 91, 91, 85]
          axes_mapping: [X, V, Z, Y, T]
          volts_per_micron: None
          axes_channels: None
          max: 5
          min: 0
```

**Review by zacsimile — APPROVED** [review]

**Pull request merged**